texastribune/dot
# Add automated way of building a new tracking script per release

#### What's this PR do?
Adds an NPM task for creating new GitHub releases. And makes it so that, on every release, a new version of the tracking script is built and placed in version control.

#### Why are we doing this? How does it help us?
The toughest part about this app is that we're releasing a script into the wild. If, down the line, we change what we track or how we want to track it, we can't also go and change the `<script>` in every republisher's CMS.

Given this, I'm making it so every release of this app also produces a tracking script with a corresponding version number. The version of the script that a republisher is using will be sent as a query parameter in the pixel-image request (and then validated via JWT). This means that, on the back end, we have the option to process requests differently based on the version number (I hope we never have to do that, but I like having the option). It also means we can 404 or redirect specific tracking requests, specifically ones coming from super old versions of our script.

#### How should this be manually tested?
+ In GitHub, [create a personal access token](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line) with `repo` scopes. Copy it. This is necessary since you'll be creating a GH release from the command line and thus accessing the API. You can delete the token right after this PR review.
+ `make -f Makefile.dev`
+ In a non-Docker terminal window that's `cd`'d to this project's root, do `npm i` and then `export GITHUB_TOKEN=<token you just got>`
+ In the same non-Docker terminal, do `npm run release`. You'll be run through a series of steps. Choose `premajor (2.0.0-0)` for the version, and answer "Y" to all other questions. Confirm it succeeds.
+ The following file should now exist: `analytics/v2.0.0-0/dot.js`
+ Go to GitHub and confirm the tag `2.0.0-0` exists and that a draft release has been created.
+ Everything worked, yay! Delete the local tag you just created (`git tag --delete 2.0.0-0`) and the remote one (`git push --delete origin 2.0.0-0`). Also delete the draft release on GH.

#### How should this change be communicated to end users?
NA.

#### Are there any smells or added technical debt to note?
It's sort of weird that Git is, like, the only thing we don't run inside Docker. But I don't really want to go thru the process of setting up Git inside a container.

#### What are the relevant tickets?
https://3.basecamp.com/3098728/buckets/736178/todos/2585881007

#### Have you done the following, if applicable:
